pinot-broker/src/main/java/com/linkedin/pinot/broker/cache/TimedRequestBrokerCache.java

```diff
@@ -0,0 +1,136 @@
+
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+import com.linkedin.pinot.common.metrics.BrokerMeter;
+import com.linkedin.pinot.common.metrics.BrokerMetrics;
+import com.linkedin.pinot.common.metrics.BrokerQueryPhase;
+import com.linkedin.pinot.common.request.BrokerRequest;
+import com.linkedin.pinot.common.response.BrokerResponse;
+import com.linkedin.pinot.pql.parsers.Pql2Compiler;
+import com.linkedin.pinot.requestHandler.BrokerRequestHandler;
+
+public class TimedRequestBrokerCache implements BrokerCache {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(TimedRequestBrokerCache.class);
+
+  private static final Pql2Compiler REQUEST_COMPILER = new Pql2Compiler();
+
+  public static String EXPIRE_TIME_VALUE = "expireTimeValue";
+  public static String EXPIRE_TIME_UNIT = "expireTimeUnit";
+  public static String BUFFER_SIZE = "bufferSize";
+
+  private CacheLoader<String, BrokerResponse> loader;
+  private LoadingCache<String, BrokerResponse> cache;
+  private BrokerRequestHandler requestHandler;
+  private BrokerMetrics brokerMetrics;
+  private int bufferSize;
+  private long expireTimeValue;
+  private TimeUnit expireTimeUnit;
+  private static ScheduledExecutorService scheduledExecutorService =
+      Executors.newSingleThreadScheduledExecutor();
+
+  public TimedRequestBrokerCache() {
+    LOGGER.info("Called constructor for TimedRequestBrokerCache!");
+  }
+
+  public void init(BrokerRequestHandler requestHandler, BrokerCacheConfigs brokerCacheConfigs,
+      BrokerMetrics brokerMetrics) {
+    this.requestHandler = requestHandler;
+    this.brokerMetrics = brokerMetrics;
+    this.loader = new ResultCacheLoader(this.requestHandler, this.brokerMetrics);
+    Map<String, String> brokerCacheClassConfigs =
+        brokerCacheConfigs.getConfigsFor(this.getClass().getSimpleName());
+    LOGGER.info("Trying to init TimedRequestBrokerCache with configs:\n{}",
+        Arrays.asList(brokerCacheConfigs));
+    this.expireTimeValue = Long.valueOf(brokerCacheClassConfigs.get(EXPIRE_TIME_VALUE));
+    this.expireTimeUnit = TimeUnit.valueOf(brokerCacheClassConfigs.get(EXPIRE_TIME_UNIT));
+    this.bufferSize = Integer.valueOf(brokerCacheClassConfigs.get(BUFFER_SIZE));
+    this.cache = CacheBuilder.newBuilder()
+        .initialCapacity(bufferSize)
+        .maximumSize(bufferSize)
+        .expireAfterWrite(expireTimeValue, expireTimeUnit)
+        .build(loader);
+    scheduledExecutorService.scheduleAtFixedRate(new Runnable() {
+      @Override
+      public void run() {
+        cache.cleanUp();
+      }
+    }, expireTimeValue, expireTimeValue, expireTimeUnit);
+    LOGGER.info("Initialized TimedRequestBrokerCache with expire time {} {} and buffer size {}",
+        this.expireTimeValue, this.expireTimeUnit, this.bufferSize);
+  }
+
+  public BrokerResponse handleRequest(JSONObject brokerRequestJson) throws Exception {
```

Comment:
I think the logic should be compile the request, check the time input in the query. If there is a start but no end, we cannot use cache. There might be other scenarios we cannot use cache. so have a method that check if a request meets the criteria and only then you go through the path of updating cache